lib/ghtorrent/ghtorrent.rb

```diff
@@ -559,21 +563,24 @@ def ensure_repo(user, repo, recursive = false)
         warn "Could not find user #{user}"
         return
       end
+      
+      #          New functionality
+      # Always persist updated repositories.  
+      # etag functionality will be used to ensure we don't
+      # hit the repo api too often
+      r = persist_repo(user, repo)
```

Comment:
This could be removed if the existing `upsert` method is re-worked to perform a MongDB update instead of a delete and insert. Otherwise, this is really driven by the need to have a MongoDB update operation.

If `upsert` is modified as described, then this call can revert back to `retrieve_repo`.

Therefore, this really depends upon how the comments above are resolved.

Reply:
I understand that every time you call this it, this will check the `etag` and fetch newer repo info, if necessary. But the problem is that `ensure_repo` is used allover the place. For example, to retrieve a PR and associated commits, `ensure_repo` will be called at least 20 times. This will cause 20 queries against the `etags` + 20 updates of the `used_count` field. If you do this in parallel with 40 clients, it will cause a massive load to DB. Hence, it is problematic.

I do not see how `upsert` (new version, as discussed above) will solve this issue, but I am curious to try. If nothing else, the existing code generates A TON of events in the streamer.